src/linker/Linker/MemberActionStore.cs

```diff
@@ -0,0 +1,83 @@
+using System.Collections.Generic;
+using Mono.Cecil;
+
+namespace Mono.Linker
+{
+	public class MemberActionStore
+	{
+		readonly Dictionary<MethodDefinition, MethodAction> _methodActions;
+		readonly Dictionary<MethodDefinition, object> _methodStubValues;
+
+		readonly Dictionary<FieldDefinition, object> _fieldValues;
+		readonly HashSet<FieldDefinition> _fieldInit;
+		readonly HashSet<AssemblyDefinition> _processedSubstitutionXml;
+		readonly LinkContext _context;
+
+		public MemberActionStore (LinkContext context)
+		{
+			_methodActions = new Dictionary<MethodDefinition, MethodAction> ();
+			_methodStubValues = new Dictionary<MethodDefinition, object> ();
+			_fieldValues = new Dictionary<FieldDefinition, object> ();
+			_fieldInit = new HashSet<FieldDefinition> ();
+			_processedSubstitutionXml = new HashSet<AssemblyDefinition> ();
+			_context = context;
+		}
+
+		void EnsureProcessedSubstitutionXml (AssemblyDefinition assembly)
+		{
+			if (_processedSubstitutionXml.Add (assembly))
+				EmbeddedXmlInfo.ProcessSubstitutions (assembly, _context);
+		}
+
+		public void SetAction (MethodDefinition method, MethodAction action)
+		{
+			_methodActions[method] = action;
+		}
+
+		public MethodAction GetAction (MethodDefinition method)
+		{
+			EnsureProcessedSubstitutionXml (method.Module.Assembly);
```

Comment:
The process-per-assembly only runs on marked assemblies, but substitutions are needed on potentially all assemblies. The constant propagation operates on any assembly it needs to - and currently it can be a loaded but unmarked assembly. It also needs this information before the assembly will be marked typically.

In theory it would be possible to integrate constant propagation into marking itself (method body processing), in which case it might be possible to only rely on marked assemblies in this case and remove this requirement - but that's LOT of refactoring and I'm still not sure it's possible to do it that way 100%.

Reply:
Alternatively we could do this for all loaded assemblies at the point where they are loaded, but that feels unnecessary.
Note that linker attributes has similar behavior - it makes sure the XML is parsed when anything asks for an attribute on a given assembly.

Reply:
> Alternatively we could do this for all loaded assemblies

That's what I meant. It seems to be much more logical than calling it within all kinds of GetXXX methods and pay extra cache hit checks everywhere.

Reply:
For substitutions it would probably work, for linker attributes it's a bit more complex as that can cause loading of additional assemblies. Doable... I guess it's a matter of preference. The current solution is as lazy as we can get. We could switch to a more eager solution.

Reply:
One principle I tried to follow is that assembly load shouldn't implicitly change global state (like the method actions here) - instead the extra processing should be done lazily by the logic that depends on it. I find this easier to reason about, but doing it lazily does require an extra check. On the other hand it can save us from doing unnecessary work for all loaded assemblies.

The different XMLs are needed at different points: it would be incorrect to process descriptor XML on assembly load, so we have to do at least that part lazily (on mark). While we *could* process attribute/substitution XML on load without introducing correctness issues, I find it more consistent to process all of the XMLs as-needed. Doing it on load would leave extra unused global state when SweepStep loads assemblies.

Are you more concerned about the lazy architecture, or about the perf impact of the check?

Reply:
To summarize our discussion - the main concern was that as written, getting attributes/actions could implicitly trigger processing which would add more attributes to the global store, and the calls to `EnsureProcessed` are easy to get wrong. I've changed it so that:
- There is an explicit difference between the global store and the per-assembly embedded XML for both substitutions and attributes
- The getters won't implicitly modify the global store, and will do XML processing as part of the cache initialization for the per-assembly stores.

Reply:
BTW @vitek-karas, I ended up avoiding structs for the `AttributeInfo` and `SubstitutionInfo` because there's no way (AFAIK) to ensure the struct members are initialized, but we assume they are non-null in a lot of places. Other options I considered were:
- Put null checks everywhere we access the struct members. This is easier to get wrong and arguably makes the code less readable. We could use `#nullable enable`, but this requires patterns like:
```csharp
if (embeddedXml.CustomAttributes?.TryGetValue (provider, out annotations) == true)
```
(the `== true` is necessary), which wasn't received well last time I introduced it.
- Struct with lazy getters replacing the fields. It can't be a readonly struct, there's just a bit more boilerplate, and the `Add*` methods either need to access the struct members through its own getters or have null checks everywhere, but it works otherwise.